core/field.ts

```diff
@@ -1015,43 +1027,50 @@ export abstract class Field<T = any> implements IASTNodeLocationSvg,
    * @returns New value, or an Error object.
    */
   private processValidation_(
-      newValue: AnyDuringMigration,
-      validatedValue: AnyDuringMigration): AnyDuringMigration {
+      newValue: AnyDuringMigration, validatedValue: T|null|undefined): T|Error {
     if (validatedValue === null) {
       this.doValueInvalid_(newValue);
       if (this.isDirty_) {
         this.forceRerender();
       }
       return Error();
     }
-    if (validatedValue !== undefined) {
-      newValue = validatedValue;
-    }
-    return newValue;
+    return validatedValue === undefined ? newValue as T : validatedValue;
   }
 
   /**
    * Get the current value of the field.
    *
    * @returns Current value.
    */
-  getValue(): AnyDuringMigration {
+  getValue(): T|null {
     return this.value_;
   }
 
   /**
-   * Used to validate a value. Returns input by default. Can be overridden by
-   * subclasses, see FieldDropdown.
+   * Validate the changes to a field's value before they are set. See
+   * **FieldDropdown** for an example of subclass implementation.
+   *
+   * **NOTE:** Validation returns one option between `T`, `null`, and
+   * `undefined`. **Field**'s implementation will never return `undefined`, but
+   * it is valid for a subclass to return `undefined` if the new value is
+   * compatible with `T`.
    *
-   * @param opt_newValue The value to be validated.
-   * @returns The validated value, same as input by default.
+   * @see {@link https://developers.google.com/blockly/guides/create-custom-blocks/fields/validators#return_values}
+   * @param newValue The value to be validated.
+   * @returns `T` to set this function's returned value instead of `newValue`.
```

Comment:
Using multiple return tags doesn't seem to be supported by the TsDoc parser and will not work with API Extractor. Could you convert this to a paragraph? You should be able to use markdown bullets if you want. I tested the parser in the playground: https://tsdoc.org/play

Reply:
Thank you for the playground link! Looks like the bulleted lists don't transfer over to the documentation, even though some other Markdown markup is. Adding an extra line between each item works decently for the documentation, though, and still appears as a bullet list in VS Code's preview of the function, so that's what I'll go with.

Let me know what you think.